### ISSUE TITLE
[UI] expanded scopes aren't updated when editing a client

### ISSUE THREAD

**Issue by djmitche** [taskcluster/taskcluster#2021]
**Describe the bug**
When a client is saved, its expandedScopes should be updated.

**To Reproduce**

1. Create a client that assumes some role
2. Edit the client to not assume the role
3. Click save

**Expected behavior**
expandedScopes updates

(I suspect the role editor has the same issue, but haven't verified)

**Comment by theprover97** [comment]
Hi @djmitche , I would like to contribute on this.

**Comment by djmitche** [comment]
Go ahead!

**Comment by theprover97** [comment]
I am not sure as how to assign role to a client. Can you tell me how to do it, so that I can recreate this bug...

**Comment by djmitche** [comment]
Hm, this might be difficult to accomplish without extra scopes assigned to you while logged in.  I've invited you to the "smoketesters" team, and once you accept you should see some scopes beginning with "auth:create-role:" and ending in "*" in https://community-tc.services.mozilla.com/profile.

Once that's in place, you can create roles and scopes that match the patterns you see in that view.

**Comment by theprover97** [comment]
I am somewhat new here, so could you tell me how to access this invite link.

**Comment by djmitche** [comment]
Hm, I'm not sure -- I would have expected GitHub to send you an email.  I re-invited you just now.  You might also see the invitation at https://github.com/orgs/taskcluster/teams/smoketesters?

**Comment by theprover97** [comment]
Thanks for the help. I am able to create clients with roles now but am facing issues regarding creating and editing roles, or assigning new roles to a client. The roles are being assigned as per scopes already defined  for smoketesters, but it is not clear as how to create entirely new roles or edit existing roles. I tried reading the documentation, but I am not able to follow it. Could you help me out?

**Comment by djmitche** [comment]
I don't think you need to create roles -- just add scopes like `assume:project:taskcluster:smoketest:fBXBgg8tRsK80ODw9bXlOQ:foo` to the client, and then remove that scope again, and you should see that the expandedRoles do not update accordinglyt.

**Comment by theprover97** [comment]
Hey there @djmitche . Sorry for this long period of inactivity. I am working on this issue, and have following points to make:-
- Refreshing the page after editing and saving the scopes changes the expanded scopes, but the same does not happen upon only clicking the ' Save Client ' button.
- The same thing is also happening with the 'Save Client' button. Saving the client after making changes does not turn the button back to grey(remains green). But it also changes properly after refreshing.
I think that the ' onclick ' event does send request at the proper time, but the response is not accounted for in its definition, hence we observe these errors.
Could you please tell me about how should these responses be handled?

**Comment by djmitche** [comment]
The first point sounds like you've correctly reproduced this, and the second point sounds like another bug, which you are also welcome to fix.

I would guess that the "onclick" is sending the request, as the client is updated -- what's missing is updating the expanded scopes based on the response to that request.

I'd like to add one additional wrinkle here: once the scopes has been edited, the expandedScopes are invalid.  Perhaps it makes sense to dynamically update that field as the scopes are edited

 * Load client page -- scopes are editable, expanded scopes are shown
 * Begin editing scopes -- expanded scopes are dynamically updated (using a technique similar to [Scope Expander](https://community-tc.services.mozilla.com/auth/scopes/expansions))
 * Click "Save Client"
 * On response, expanded scopes are updated based on the result of updateClient operation.
